src/core-layers/grid-layer/grid-layer.js

```diff
@@ -56,114 +69,238 @@ export default class GridLayer extends CompositeLayer {
+    // will automatically be called
+    return {
+      getColor: [
+        {
+          id: 'value',
+          triggers: ['getColorValue'],
+          updater: this.getSortedColorBins
+        }, {
+          id: 'domain',
+          triggers: ['lowerPercentile', 'upperPercentile'],
+          updater: this.getColorValueDomain
+        }, {
+          id: 'scaleFunc',
+          triggers: ['colorDomain', 'colorRange'],
+          updater: this.getColorScale
+        }
+      ],
+      getElevation: [
+        {
+          id: 'value',
+          triggers: ['getElevationValue'],
+          updater: this.getSortedElevationBins
+        }, {
+          id: 'domain',
+          triggers: ['elevationLowerPercentile', 'elevationUpperPercentile'],
+          updater: this.getElevationValueDomain
+        }, {
+          id: 'scaleFunc',
+          triggers: ['elevationDomain', 'elevationRange'],
+          updater: this.getElevationScale
+        }
+      ]
+    };
+  }
 
-    } else if (this.needsRecalculateColorDomain(oldProps, props)) {
+  getDimensionChanges(oldProps, props) {
+    const {dimensionUpdaters} = this.state;
+    // dimension should be updated
 
-      this.getValueDomain();
-    }
-  }
+    const updaters = Object.keys(dimensionUpdaters).reduce((accu, key) => {
 
-  needsReProjectPoints(oldProps, props) {
-    return oldProps.cellSize !== props.cellSize;
+      // return the first triggered updater for each dimension
+      const updater = dimensionUpdaters[key]
+        .find(item => item.triggers.some(t => oldProps[t] !== props[t]));
+
+      if (updater) {
+        accu.push(updater.updater);
+      }
+
+      return accu;
+    }, []);
+
+    return updaters.length ? updaters : null;
   }
 
-  needsRecalculateColorDomain(oldProps, props) {
-    return oldProps.lowerPercentile !== props.lowerPercentile ||
-      oldProps.upperPercentile !== props.upperPercentile;
+  getPickingInfo({info}) {
+    const {sortedColorBins, sortedElevationBins} = this.state;
+
+    const isPicked = info.picked && info.index > -1;
+    let object = null;
+
+    if (isPicked) {
+      const cell = this.state.layerData[info.index];
+
+      const colorValue = sortedColorBins.binMap[cell.index] &&
+        sortedColorBins.binMap[cell.index].value;
+      const elevationValue = sortedElevationBins.binMap[cell.index] &&
+        sortedElevationBins.binMap[cell.index].value;
+
+      object = Object.assign({
+        colorValue,
+        elevationValue
+      }, cell);
+    }
+
+    // add bin colorValue and elevationValue to info
+    return Object.assign(info, {
+      picked: Boolean(object),
+      // override object with picked cell
+      object
+    });
   }
 
-  needsReSortBins(oldProps, props) {
-    return oldProps.getColorValue !== props.getColorValue;
+  getUpdateTriggers() {
+    const {dimensionUpdaters} = this.state;
+    // merge all dimension triggers
+    return Object.keys(dimensionUpdaters).reduce((accu, dimension) => {
```

Comment:
Nit: This is basically just opinions, so take them with a grain of salt:

I find this code is hard to follow - partly because I am personally not a fan of `reduce` calls, in spite of their theoretical appeal I find them hard to read, especially the JavaScript `reduce` syntax.

But more importantly, even if this was a `for of loop`, again it feels like this type of generic iteration code would belong in a generic aggregation manager, where it would make perfect sense, but in a layer I would rather see code that explicitly treated each attribute, and kept things concrete. 

Put another way: Are we really saving a lot of typing by raising the abstraction level like this?